### PR TITLE
(ANN) Expand syntax hightlighting to more types.

### PR DESCRIPTION
**Pull request by mfarrugi**
Remade #503 

Unsquashed commits visible at [mfarrugi:SyntaxHighlighting](/mfarrugi/intellij-rust/tree/SyntaxHighlighting)

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/annotator/RustHighlightingAnnotator.kt [hunk 1]
```diff
@@ -2,73 +2,89 @@ package org.rust.ide.annotator
...
+    }
+
+    fun AnnotationHolder.highlight(element: PsiElement?, color: RustColor?) {
+        val textAttributesKey = color?.textAttributesKey
+        if (element != null && textAttributesKey != null) {
+            createInfoAnnotation(element, null).textAttributes = textAttributesKey
         }
+    }
+
+    fun highlightingVisitor() = object : RustComputingVisitor<Pair<PsiElement?, RustColor?>>() {
```
Comment (jajakobyly) [inline src/main/kotlin/org/rust/ide/annotator/RustHighlightingAnnotator.kt]: Why not just:

```kotlin
private class HighlightingVisitor : RustComputingVisitor<Pair<PsiElement?, RustColor?>>() { /* ... */ }
```

File: src/main/kotlin/org/rust/ide/annotator/RustHighlightingAnnotator.kt [hunk 2]
```diff
@@ -2,73 +2,89 @@ package org.rust.ide.annotator
...
-            val color = if (o.isStatic) RustColor.STATIC_METHOD else RustColor.INSTANCE_METHOD
-            holder.highlight(o.identifier, color)
-        }
-    })
+        override fun visitExternCrateItem(o: RustExternCrateItemElement) = highlight(o.identifier, RustColor.CRATE)
 
-    private fun AnnotationHolder.highlight(element: PsiElement?, color: RustColor?) {
-        highlight(element, color?.textAttributesKey)
-    }
+        //TODO: When are element and element.identifier different?
```
Comment (jajakobyly) [inline src/main/kotlin/org/rust/ide/annotator/RustHighlightingAnnotator.kt]: Macro invocation's production is:
```
elementType (".*_macro_invocation|macro_rules_invocation") = macro_invocation

...

// Invocations
macro_invocation ::= IDENTIFIER '!'
try_macro_invocation ::= "try" '!'
macro_rules_invocation ::= "macro_rules" '!'

// https://doc.rust-lang.org/std/fmt/
format_like_macro_invocation ::= ( "format"
                                 | "format_args"
                                 | "write"
                                 | "writeln"
                                 | "print"
                                 | "println"
                                 // logging macros from the log crate
                                 | "debug"
                                 | "error"
                                 | "info"
                                 | "log"
                                 | "trace"
                                 | "warn" ) '!'
```

So as you can see, `element.identifier` is the part without `!` here

**Comment by matklad** [comment]
Thanks! I've merge this manually because it had conflicts with the markdown lexing pull request.

**Comment by matklad** [comment]
CC @himikof, because it is the highlighter.

**Pull request closed without merging**